src/PushCommand.php

```diff
@@ -305,17 +316,42 @@ private function sendFile(
      *
      * @param $url
      * @param $file
+     * @param $sourceType
+     * @param $sourceUrl
+     * @param $sourceReference
      * @param $username
      * @param $password
      *
      * @throws \GuzzleHttp\Exception\GuzzleException
      */
-    private function postFile($url, $file, $username = null, $password = null)
+    private function postFile($url, $file, $sourceType = null, $sourceUrl = null, $sourceReference = null, $username = null, $password = null)
     {
         $options = [
-            'body' => fopen($file, 'r'),
             'debug' => $this->getIO()->isVeryVerbose(),
         ];
+        if (!empty($sourceType) && !empty($sourceUrl) && !empty($sourceReference)) {
```

Comment:
Is it possible to have only a subset of those options specified? Currently, if you set two out of three, everything is ignore.
An other point: does it makes sens to put it also as a parameter in `composer.json`?

Reply:
It would also be lovely to specify only a flag to enable the source tag and maybe a kind of auto-sense for looking if there is a .git directory and automatically detecting the upstream origin url. It would also be possible to have all of the auto-sense values being specified in the composer.json.

Reply:
> Is it possible to have only a subset of those options specified? Currently, if you set two out of three, everything is ignore.

It doesn't make sense to continue with just 2 of 3. But I think it would be better to inform the user about his misconfiguration instead of just uploading the file alone.

Reply:
I think informing the user about his misconfiguration is a good idea, it will avoid bad surprises.
Can you update the code accordingly @thirsch?
Thanks!

Reply:
Hey guys, sorry for the delay, I was really busy with delivery of our project to the client.
I'll check your request and will do so today ;)

Reply:
> It would also be lovely to specify only a flag to enable the source tag and maybe a kind of auto-sense for looking if there is a .git directory and automatically detecting the upstream origin url. It would also be possible to have all of the auto-sense values being specified in the composer.json.

Well detecting git is not a problem, but determining which upstream should be used, maybe problematic as we may have multiple, and maybe not willing to expose the internal one, but the external (public one).
The reason that I didn't add any auto detection on the process was giving the possibility to the user to do so, not to be intrusive. There are couple of simple git commands that can be used to detect these and that's the way my company is using this plugin right now.
In our case:
`
  --src-type=git
  --src-url="$(git remote get-url origin)"
  --src-ref="$(git rev-parse HEAD)"
`
In the publish process.
Maybe having a new option such as --src-git as an optional which uses these three options internally can be a good idea.
What do you think?

Reply:
I just pushed the modification for the verification of all or none for the parameters.
Do let me know if you think anything else would be useful, but not intrusive.
And the reason that I didn't add anything in the composer.json file was that maybe you would like to have different versions on different remote servers. So the URL can't be there.
The reference is definitely not possible to be on composer.json and the source type is the only thing that may have sense but as it can't do anything on it's own, well, not that interesting as well.

Reply:
From my point of view, it was all that's needed to get this pr merged. Everything else is nice to have and can be proposed by upcoming prs?

Reply:
Great, and thanks, so I would be glad if the PR can be merged and I will switch our app to the official version.